Move BadHttpRequestException to Http.Abstractions

cado1982 opened the pull request:
Summary of the changes (Less than 80 chars)
 - Replace Kestrel and IIS versions of BadHttpRequestException with shared version in Http Abstractions

Addresses #11208 

This is my first contribution so I'm open to any feedback you may have.

cado1982 committed:
Add new BadHttpRequestException to Http Abstractions

cado1982 committed:
Update Kestrel to use new BadHttpRequestException from Http Abstractions

cado1982 committed:
Update IIS to use new BadHttpRequestException from Http Abstractions

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=20339) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=20339) before we can accept your contribution.<br/><br/>:x: cado1982 [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=20339)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=20339) it.</sub>

cado1982 committed:
Add missing using directive to qualify StatusCodes

cado1982 committed:
Regenerate reference assemblies

Tratcher commented:
@halter73 @anurse https://github.com/dotnet/aspnetcore/issues/15038 proposed making the old BadHttpRequestException derive from the new one to limit the breaking change impact. The servers would continue throwing the old ones for now.

https://github.com/dotnet/aspnetcore/issues/17804 also proposed making the Reason enum public for more granular error management.

These had both been resolved as duplicates of #11208 but the context didn't get pulled over.

halter73 commented on src/Servers/Kestrel/Core/src/BadHttpRequestException.cs:
```diff
@@ -11,6 +12,7 @@
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core
 {
+    [Obsolete("Moved to Microsoft.AspNetCore.Http.BadHttpRequestException")]
```
As @Tratcher said, we should make this and the IIS BadHttpRequestException derive from Microsoft.AspNetCore.Http.BadHttpRequestException and we should continue throwing the more-derived types for max back-compat. Sorry I forgot to mention this earlier.

I think we can look at making the Reason enum public in a separate PR. I'm not sure the entire enum should be public. A lot of the Reasons are internal to Kestrel and would never be set on an exception that could be caught by the application.

halter73 commented on src/Servers/Kestrel/Core/src/KestrelBadHttpRequestException.cs:
```diff
@@ -0,0 +1,177 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core
+{
+    internal class KestrelBadHttpRequestException : IOException
```
It might be easier to have this derive from Kestrel's BadHttpRequestException.

Tratcher replied:
I'm not sure why this copy of BadHttpRequestException even exists?

halter73 replied:
That's true, if we're going to be throwing the old type anyway, we can just keep the internal stuff there for now.

anurse commented:
I reopened the Reason one, we can do that in a separate PR.

cado1982 commented:
@halter73 Let me recap what I think I need to do. Correct me if any points are wrong.

- Keep the newly created Microsoft.AspNetCore.Http.BadHttpRequestException but it won't be sealed now
- Make IIS and Kestrel's BadHttpRequestException derive from Microsoft.AspNetCore.Http.BadHttpRequestException
- Remove the Obsolete attributes from IIS and Kestrel's old BadHttpRequestException, they'll still be used as they are now
- Move all of the internal members from Kestrel's old BadHttpRequestException and have them in KestrelBadHttpRequestException instead which will derive from Kestrel's old BadHttpRequestException 

I also have a question about AllowedHeader. That is currently only used in the following two places https://github.com/dotnet/aspnetcore/blob/d1d9b97f77cfe672618512e30b5b78b81ce44ff2/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L427 
https://github.com/dotnet/aspnetcore/blob/d1d9b97f77cfe672618512e30b5b78b81ce44ff2/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L471

The property is then used to set the http header here 

https://github.com/dotnet/aspnetcore/blob/d1d9b97f77cfe672618512e30b5b78b81ce44ff2/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L1263

Can the response header be set in the place the exceptions are thrown in Http1Connection or is it too early in the request pipeline?

halter73 commented:
@cado1982 Looks correct except one thing:

> Remove the Obsolete attributes from IIS and Kestrel's old BadHttpRequestException, they'll still be used as they are now

Keep obsoleting IIS's and Kestrel's BadHttpRequestException. You'll probably need add some pragmas to silence warnings about our own internal usage, but we want to encourage customers to catch the base type. Eventually we might remove the obsoleted derived types or make them internal.

cado1982 committed:
Revert c7a356 f81b48 7b5611 aaea8b

cado1982 committed:
Fix merge conflicts

cado1982 committed:
Alter BadHttpRequestException to inherit from shared Http Abstractions version

- Mark existing BadHttpRequestException in IIS and Kestrel Obsolete
- Add pragmas to ignore internal Obsolete usage

cado1982 committed:
Regenerate reference assemblies

cado1982 commented:
@halter73 @anurse @Tratcher I'm not sure about this failing build. Did I break something?

Tratcher commented:
> @halter73 @anurse @Tratcher I'm not sure about this failing build. Did I break something?

No, looks like a networking issue. I've re-run it.

Tratcher commented on src/Servers/IIS/IIS/src/BadHttpRequestException.cs:
```diff
@@ -1,23 +1,22 @@
             Reason = reason;
         }
 
-        public int StatusCode { get; }
```
How concerned are we about the binary breaking change here? It could be mitigated with:
`public new int StatusCode { get => base.StatusCode }`

Tratcher commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:
```diff
@@ -122,7 +122,9 @@ private async Task ReadBody()
                     if (_consumedBytes > MaxRequestBodySize)
                     {
+#pragma warning disable CS0618 // Type or member is obsolete
                         BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTooLarge);
```
Lets move the static Throw API to a non obsolete internal type so we don't have to use pragma everywhere. I know that's what you were trying to do with the KestrelBadHttpRequestException, but I think we only need to move the static API.

Tratcher commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.Log.cs:
```diff
@@ -38,7 +38,9 @@ public static void UnexpectedError(ILogger logger, string className, Exception e
             }
 
+#pragma warning disable CS0618 // Type or member is obsolete
             public static void ConnectionBadRequest(ILogger logger, string connectionId, BadHttpRequestException ex)
```
Most of these usages can switch to referencing the base type either inline or with a using statement in the file. That will avoid excess pragmas. The two places we need to directly reference the old types are when it's first created & thrown, as well as tests that check which type was thrown. All of the other catch blocks and fields can use the new base type.

Tratcher commented on src/Http/Http.Abstractions/src/BadHttpRequestException.cs:
```diff
@@ -0,0 +1,36 @@
+        /// </summary>
+        /// <param name="message">The message to associate with this exception.</param>
+        /// <param name="statusCode">The HTTP status code to associate with this exception.</param>
+        public BadHttpRequestException(string message, int statusCode)
```
Please add an overload that accepts an inner exception.

Tratcher commented on src/Servers/Kestrel/Core/src/BadHttpRequestException.cs:
```diff
@@ -29,8 +30,6 @@ private BadHttpRequestException(string message, int statusCode, RequestRejection
             }
         }
 
-        public int StatusCode { get; }
```
Same issue here: `public new int StatusCode { get => base.StatusCode; }`

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:
```diff
@@ -150,7 +150,9 @@ public void StopProcessingNextRequest()
         public void HandleRequestHeadersTimeout()
         {
+#pragma warning disable CS0618 // Type or member is obsolete
             Log.ConnectionBadRequest(ConnectionId, BadHttpRequestException.GetException(RequestRejectionReason.RequestHeadersTimeout));
```
Ah, GetException should move to a new static type as well.

Tratcher commented on src/Servers/Kestrel/Core/test/HttpParserTests.cs:
```diff
@@ -90,11 +90,15 @@ public void ParseRequestLineThrowsOnInvalidRequestLine(string requestLine)
             var requestHandler = new RequestHandler();
 
+#pragma warning disable CS0618 // Type or member is obsolete
             var exception = Assert.Throws<BadHttpRequestException>(() =>
```
This is a good example of where we want to keep the pragmas so we can test that we're throwing the old exception type.

Tratcher reviewed (commented):
Looks pretty good. Most my comments are around reducing the need for pragmas.

cado1982 committed:
Further refactorings for IIS and Kestrel BadHttpRequestException

- StatusCode now simply returns the base.StatusCode
- Static API from IIS and Kestrel BadHttpRequestException moved to respective IISBadHttpRequestException and KestrelBadHttpRequestException types
- Most usages of deprecated IIS and Kestrel BadHttpRequestException moved to base type
- Overload added to Microsoft.AspNetCore.Http.BadHttpRequestException to accept inner exception

cado1982 committed:
Regenerate reference assemblies

Tratcher commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:
```diff
@@ -26,6 +26,8 @@
 
 namespace Microsoft.AspNetCore.Server.IIS.Core
 {
+    using BadHttpRequestException = Microsoft.AspNetCore.Http.BadHttpRequestException;
```
Style: Usings go outside the namespace where possible.

cado1982 replied:
It's not possible here. Placing the using outside of the namespace still uses the Obsolete version of BadHttpRequestException.

Tratcher replied:
Oh, because the obsolete one is in the Microsoft.AspNetCore.Server.IIS.Core namespace?

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs:
```diff
@@ -52,7 +52,7 @@ protected override Task OnConsumeAsync()
                 }
             }
-            catch (BadHttpRequestException ex)
+            catch (Microsoft.AspNetCore.Http.BadHttpRequestException ex)
```
Using statement?

cado1982 replied:
I can use a using directive here but it must go inside the namespace again.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:
```diff
@@ -450,8 +451,10 @@ private unsafe void RejectUnknownVersion(byte* version, int length)
 
         [MethodImpl(MethodImplOptions.NoInlining)]
+#pragma warning disable CS0618 // Type or member is obsolete
         private unsafe BadHttpRequestException GetInvalidRequestException(RequestRejectionReason reason, byte* detail, int length)
```
Use base type / using statement?

cado1982 replied:
Callers of this method throw the returned exception so I think it needs to be the old obsoleted version?

Tratcher replied:
Only the statement `new BadHttpRequestException` needs to specifically use the old type. All other fields, parameters, and return values can use the base type, they're just passing it through.

cado1982 replied:
Learnt something here. No problem, fixing.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -1313,8 +1317,10 @@ public void ThrowRequestTargetRejected(Span<byte> target)
-        private BadHttpRequestException GetInvalidRequestTargetException(Span<byte> target)
-            => BadHttpRequestException.GetException(
+#pragma warning disable CS0618 // Type or member is obsolete
+        private Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException GetInvalidRequestTargetException(Span<byte> target)
```
Base type/using?

cado1982 replied:
Again callers of this method throw the returned exception so I think it needs to be the old obsoleted version?

Tratcher reviewed (approved):
Much better.

cado1982 committed:
Fix some usages of BadHttpRequestException and add using statements

Tratcher commented:
Looks like a test issue in IIS.Tests.MaxRequestBodySizeTests.EveryReadFailsWhenContentLengthHeaderExceedsGlobalLimit:
https://dev.azure.com/dnceng/public/_build/results?buildId=587168&view=ms.vss-test-web.build-test-results-tab&runId=18503624&resultId=119834&paneView=debug

```
Assert.Equal() Failure

↓ (pos 9)

Expected: HTTP/1.1 413 Payload Too Large

Actual: HTTP/1.1 500 Internal Server E

↑ (pos 9)
```

I think it's this part:
https://github.com/dotnet/aspnetcore/pull/20339/files#diff-f7fe79bd73501e55cf962d965d1dce28R320-R323
```
                    requestRejectedEx1 = await Assert.ThrowsAsync<BadHttpRequestException>(
                        async () => await ctx.Request.Body.ReadAsync(buffer, 0, 1));
                    requestRejectedEx2 = await Assert.ThrowsAsync<BadHttpRequestException>(
                        async () => await ctx.Request.Body.ReadAsync(buffer, 0, 1));
```
ThrowsAsync checks for the exact exception type, not derived types. You can use ThrowsAnyAsync or check for the derived type here.

cado1982 committed:
Merge branch 'master' into temp-11208

cado1982 committed:
Change test from ThrowsAny to ThrowsAnyAsync

cado1982 commented:
I applied the change you mentioned but I'm struggling to run the IIS tests locally. They run continuously and don't complete (for at least 2 hours). Was that the only one that was broken?

Tratcher commented:
Looks like you fixed it. The remaining issue with Node seems to be unrelated, I'll see to that.

halter73 reviewed (approved):
Some of the test #pragma's could be removed, but nbd.

Tratcher commented:
Thanks @cado1982

cado1982 commented:
@halter73 @Tratcher Thanks. I learnt a lot from this.
